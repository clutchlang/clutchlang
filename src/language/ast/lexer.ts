export * from './lexer/token';
export * from './lexer/tokenizer';
