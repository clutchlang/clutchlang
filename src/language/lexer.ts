import { ISourceSpan, StringScanner } from '../agnostic/scanner';
import {
  Characters,
  isDigit,
  isHexadecimal,
  isLetter,
  isWhiteSpace,
} from '../agnostic/strings';

function isIdentifier(character: number): boolean {
  return isIdentifierStart(character) || isDigit(character);
}

function isIdentifierStart(character: number): boolean {
  return isLetter(character) || character === Characters.$_;
}

/**
 * Tokenizes and returns @param program as a series of @see Token.
 */
export function tokenize(program: string, onError?: ErrorReporter): Token[] {
  return new ClutchLexer().scanProgram(new StringScanner(program), onError);
}

/**
 * Reports a @param message at the provided @param span.
 */
export type ErrorReporter = (span: ISourceSpan, error: string) => void;

export enum TokenKind {
  // Literals
  IDENTIFIER = 'Identifier',
  STRING = 'String',
  NUMBER = 'Number',

  // Keywords
  CLASS = 'class',
  ELSE = 'else',
  FALSE = 'false',
  FOR = 'for',
  IF = 'if',
  LET = 'let',
  RETURN = 'return',
  SUPER = 'super',
  THEN = 'then',
  THIS = 'this',
  TRUE = 'true',
  WHILE = 'while',

  // Symbols
  ARROW = '->',
  LEFT_PAREN = '(',
  RIGHT_PAREN = ')',
  LEFT_CURLY = '{',
  RIGHT_CURLY = '}',

  // Operators
  PERIOD = '.',
  PLUS = '+',
  PLUS_EQUALS = '+=',
  MINUS = '-',
  MINUS_EQUALS = '-=',
  STAR = '*',
  STAR_EQUALS = '*=',
  SLASH = '/',
  SLASH_EQUALS = '/=',
  MODULUS = '%',
  MODULUS_EQUALS = '%=',
  EQUALS = '=',
  EQUALS_EQUALS = '==',
  EXCLAIM_EQUALS = '!=',
  EQUALS_EQUALS_EQUALS = '===',
  EXCLAIM_EQUALS_EQUALS = '!==',
  RIGHT_ANGLE = '>',
  RIGHT_ANGLE_EQUALS = '>=',
  LEFT_ANGLE = '<',
  LEFT_ANGLE_EQUALS = '<=',
  PIPE = '|',
  PIPE_PIPE = '||',
  AND = '&',
  AND_AND = '&&',
  CARET = '^',
  EXCLAIM = '!',
  TILDE = '~',
  PLUS_PLUS = '++',
  MINUS_MINUS = '--',
  LEFT_ANGLE_LEFT_ANGLE = '<<',
  RIGHT_ANGLE_RIGHT_ANGLE = '>>',

  // Misc
  EOF = '<EOF>',
}

/**
 * A language-level lexer that yields tokens understood by the grammar.
 */
export class ClutchLexer {
  public static readonly keywords: {
    [index: string]: TokenKind;
  } = {
    class: TokenKind.CLASS,
    else: TokenKind.ELSE,
    false: TokenKind.FALSE,
    for: TokenKind.FOR,
    if: TokenKind.IF,
    let: TokenKind.LET,
    return: TokenKind.RETURN,
    super: TokenKind.SUPER,
    this: TokenKind.THIS,
    true: TokenKind.TRUE,
    while: TokenKind.WHILE,
  };

  private program!: StringScanner;
  private onError!: ErrorReporter;
  private position!: number;
  private lastComments: IComment[] = [];

  public scanProgram(
    program: StringScanner,
    onError: ErrorReporter = (span, message) => {
      throw new SyntaxError(
        `${message} "${span.text}" at ${span.line}:${span.column}`
      );
    }
  ): Token[] {
    this.position = 0;
    this.program = program;
    this.onError = onError;
    const tokens: Token[] = [];
    while (program.hasNext()) {
      const token = this.scanToken();
      if (token) {
        tokens.push(token);
      }
    }
    tokens.push(this.createToken(TokenKind.EOF, ''));
    return tokens;
  }

  private scanToken(): Token | undefined {
    const character = this.program.advance();
    switch (character) {
      case Characters.$LPAREN: // "("
        return this.createToken(TokenKind.LEFT_PAREN);
      case Characters.$RPAREN: // ")"
        return this.createToken(TokenKind.RIGHT_PAREN);
      case Characters.$LCURLY: // "{"
        return this.createToken(TokenKind.LEFT_CURLY);
      case Characters.$RCURLY: // "}"
        return this.createToken(TokenKind.RIGHT_CURLY);
      case Characters.$PERIOD: // "."
        return this.createToken(TokenKind.PERIOD);
      case Characters.$PLUS: // "+" or "+=" or "++"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.PLUS_EQUALS
            : this.program.match(Characters.$PLUS)
              ? TokenKind.PLUS_PLUS
              : TokenKind.PLUS
        );
      case Characters.$MINUS: // "-" or "--" or "-=" or "->"
        return this.createToken(
          this.program.match(Characters.$RANGLE)
            ? TokenKind.ARROW
            : this.program.match(Characters.$EQUALS)
              ? TokenKind.MINUS_EQUALS
              : this.program.match(Characters.$MINUS)
                ? TokenKind.MINUS_MINUS
                : TokenKind.MINUS
          // Code coverage considers the `);` to be a missed line :(
          /* istanbul ignore next */
        );
      case Characters.$STAR: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.STAR_EQUALS
            : TokenKind.STAR
        );
      case Characters.$PERCENT: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.MODULUS_EQUALS
            : TokenKind.MODULUS
        );
      case Characters.$EQUALS: // "=" or "==" or "==="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? TokenKind.EQUALS_EQUALS_EQUALS
              : TokenKind.EQUALS_EQUALS
            : TokenKind.EQUALS
        );
      case Characters.$LANGLE: // "<" or "<=" or "<<"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.LEFT_ANGLE_EQUALS
            : this.program.match(Characters.$LANGLE)
              ? TokenKind.LEFT_ANGLE_LEFT_ANGLE
              : TokenKind.LEFT_ANGLE
        );
      case Characters.$RANGLE: // ">" or ">=" or ">>"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? TokenKind.RIGHT_ANGLE_EQUALS
            : this.program.match(Characters.$RANGLE)
              ? TokenKind.RIGHT_ANGLE_RIGHT_ANGLE
              : TokenKind.RIGHT_ANGLE
        );
      case Characters.$SLASH: // "/" or "//" or "/="
        return this.scanSlash();
      case Characters.$SQUOTE: // "'"
        return this.scanString();
      case Characters.$PIPE: // "|" or "||"
        return this.createToken(
          this.program.match(Characters.$PIPE)
            ? TokenKind.PIPE_PIPE
            : TokenKind.PIPE
        );
      case Characters.$AND: // "&" or "&&"
        return this.createToken(
          this.program.match(Characters.$AND)
            ? TokenKind.AND_AND
            : TokenKind.AND
        );
      case Characters.$TILDE: // "~"
        return this.createToken(TokenKind.TILDE);
      case Characters.$CARET: // "^"
        return this.createToken(TokenKind.CARET);
      case Characters.$EXCLAIM: // "!" or "!=" or "!==""
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? TokenKind.EXCLAIM_EQUALS_EQUALS
              : TokenKind.EXCLAIM_EQUALS
            : TokenKind.EXCLAIM
        );
      default:
        if (isWhiteSpace(character)) {
          this.position = this.program.position;
          break;
        }
        if (isDigit(character)) {
          return this.scanNumber(character);
        }
        if (isIdentifierStart(character)) {
          return this.scanIdentifierOrKeyword();
        }
        this.reportError(
          `Unexpected character "${String.fromCharCode(character)}"`
        );
    }
    return;
  }

  /**
   * Scans a string, returning `undefined` if there was no string.
   */
  private scanString(): Token {
    while (this.program.hasNext()) {
      if (this.program.match(Characters.$SQUOTE)) {
        this.position++;
        let lexeme = this.substring();
        lexeme = lexeme.substring(0, lexeme.length - 1);
        return this.createToken(TokenKind.STRING, lexeme);
      }
      this.program.advance();
    }
    this.reportError('Unterminated string');
    this.position++;
    return this.createToken(TokenKind.STRING);
  }

  /**
   * Scans any tokens that are valid numeric literals
   */
  private scanNumber(starting: number): Token {
    if (starting === Characters.$0) {
      if (
        this.program.match(Characters.$X) ||
        this.program.match(Characters.$x)
      ) {
        return this.scanHexNumber();
      }
    }
    if (
      this.program.match(Characters.$E) ||
      this.program.match(Characters.$e)
    ) {
      return this.scanDigits();
    }
    while (this.program.hasNext()) {
      const next = this.program.peek();
      if (this.program.match(Characters.$PERIOD)) {
        break;
      }
      if (!isDigit(next)) {
        break;
      }
      this.program.advance();
    }
    return this.scanDigits();
  }

  /**
   * Scans any tokens that are valid hexadecimel digits, returning a number.
   */
  private scanHexNumber(): Token {
    this.scanPredicate(isHexadecimal);
    return this.createToken(TokenKind.NUMBER);
  }

  /**
   * Scans any tokens that are valid digits, returning a number.
   */
  private scanDigits(): Token {
    this.scanPredicate(isDigit);
    return this.createToken(TokenKind.NUMBER);
  }

  /**
   * Scans any tokens that are valid identifiers.
   */
  private scanIdentifierOrKeyword(): Token {
    this.scanPredicate(isIdentifier);
    const identifierOrKeyword = this.substring();
    return new Token(
      ClutchLexer.keywords[identifierOrKeyword] || TokenKind.IDENTIFIER,
      identifierOrKeyword,
      this.clearComments(),
      this.position - identifierOrKeyword.length
    );
  }

  /**
   * Scans tokens preceding with "/".
   */
  private scanSlash(): Token | undefined {
    // "//"
    if (this.program.match(Characters.$SLASH)) {
      this.advanceToEOL();
      this.lastComments.push({
        lexeme: this.substring().trim(),
        offset: this.position,
      });
      return;
    }
    // "/" or "/="
    return this.createToken(
      this.program.match(Characters.$EQUALS)
        ? TokenKind.SLASH_EQUALS
        : TokenKind.SLASH
    );
  }

  /**
   * Scans until the EOF or until @param predicate returns false.
   */
  private scanPredicate(predicate: (character: number) => boolean): void {
    while (this.program.hasNext()) {
      const peek = this.program.peek();
      if (!predicate(peek)) {
        break;
      }
      this.program.advance();
    }
  }

  /**
   * Advances the position of the scanner until either EOF or EOL.
   *
   * **NOTE**: Treats `CR+LF` as a DOS-stlye line ending, not two lines.
   */
  private advanceToEOL(): void {
    while (this.program.hasNext()) {
      // "\n" (LF)
      if (this.program.match(Characters.$LF)) {
        return;
      }
      // "\r" (CR)
      if (this.program.match(Characters.$CR)) {
        // "\r\n" (CR+LF)
        this.program.match(Characters.$LF);
        return;
      }
      this.program.advance();
    }
  }

  /**
   * Returns a new token with a @param kind and the current substring.
   */
  private createToken(kind: TokenKind, content = this.substring()): Token {
    return new Token(
      kind,
      content,
      this.clearComments(),
      this.position - content.length
    );
  }

  /**
   * Clear and retrieve comments lexed.
   */
  private clearComments(): IComment[] {
    const comments = this.lastComments;
    this.lastComments = [];
    return comments;
  }

  /**
   * Returns the current substring, advancing the position counter.
   */
  private substring(): string {
    const start = this.position;
    const end = this.program.position;
    this.position = end;
    return this.program.substring(start, end);
  }

  /**
   * Reports a lexical @param error occurring at @param offset.
   */
  private reportError(error: string, offset = this.position): void {
    this.onError(this.program.source.span(offset, offset + 1), error);
  }
}

/**
 * A scanned token from @function tokenize.
 */
export interface IToken {
  readonly kind: TokenKind;
  readonly lexeme: string;
  readonly comments: IComment[];
  readonly offset: number;
}

/**
 * Implementation of @interface IToken private to @function tokenize.
 */
class Token implements IToken {
  constructor(
    public readonly kind: TokenKind,
    public readonly lexeme: string,
    public readonly comments: IComment[],
    public readonly offset: number
  ) {}
}

export interface IComment {
  readonly lexeme: string;
  readonly offset: number;
}
