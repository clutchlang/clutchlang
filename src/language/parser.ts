export * from './parser/expressions';
export * from './parser/factory';
export * from './parser/nodes';
export * from './parser/operators';
export * from './parser/statements';
export * from './parser/visitors';
