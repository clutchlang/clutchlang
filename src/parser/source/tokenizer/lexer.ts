import { Characters } from './characters';
import { SourceScanner } from './scanner';
import {
  RegExpToken,
  StringToken,
  SymbolToken,
  Token,
  TokenKind,
} from './tokens';

/**
 * Produces a stream of tokens from the @member scanner.
 *
 * Implements @type {Iterable<Token>}.
 */
export class Lexer implements Iterable<Token> {
  /**
   * Invalid identifiers.
   */

  private static readonly keywords = new Set(['true', 'false']);

  constructor(private readonly scanner: SourceScanner) {}

  private get substring() {
    let substring = this.scanner.substring;
    const newLine = substring.indexOf('\n');
    if (newLine !== -1) {
      substring = substring.substring(0, newLine);
    }
    return substring;
  }

  public *[Symbol.iterator](): Iterator<Token> {
    this.consumeWhitespace();
    while (!this.scanner.isDone) {
      yield* this.scanCompilationUnit();
      this.consumeWhitespace();
    }
  }

  protected consumeWhitespace(): void {
    this.scanner.scan(/\s+/);
  }

  /**
   * Throws a terminal error at the current scanner position.
   *
   * @param message
   */
  protected fail(message: string): never {
    throw new SyntaxError(this.scanner.span.highlight(message));
  }

  protected scanOptional(kind: TokenKind): Token | undefined {
    this.consumeWhitespace();
    const result = this.scanner.scan(kind.pattern);
    return result
      ? new Token(kind, this.scanner.lastSpan(), this.scanner.lastMatch![0])
      : undefined;
  }

  protected scanRequired(kind: TokenKind): Token {
    this.consumeWhitespace();
    const result = this.scanner.scan(kind.pattern);
    if (!result) {
      let pattern = kind.pattern;
      /* istanbul ignore next */
      if (typeof pattern === 'number') {
        /* istanbul ignore next */
        pattern = String.fromCharCode(pattern);
      } else if (pattern instanceof RegExp) {
        /* istanbul ignore next */
        pattern = pattern.source;
      }
      this.fail(`Expected ${pattern}, got "${this.substring}"`);
    }
    return new Token(kind, this.scanner.lastSpan(), this.scanner.lastMatch![0]);
  }

  protected scanCompilationUnit(): Iterable<Token> {
    return this.scanFunction();
  }

  protected *scanExpression(): Iterable<Token> {
    const literal = this.scanLiteral();
    if (literal) {
      yield literal;
      return;
    }
    const identifier = this.scanOptional(RegExpToken.Identifier);
    if (identifier) {
      yield identifier;
    }
    if (this.scanner.peek() === Characters.LParen) {
      yield* this.scanParantheses();
    } else if (!identifier) {
      const substring = this.substring;
      this.scanner.read();
      this.consumeWhitespace();
      return this.fail(`Expected expression, got "${substring}"`);
    }
  }

  protected *scanExpressionOrBlock(): Iterable<Token> {
    this.consumeWhitespace();
    if (this.scanner.peek() === Characters.LCurly) {
      yield this.scanRequired(SymbolToken.LCurly);
      let checkRCurly = true;
      while (
        !this.scanner.isDone &&
        this.scanner.peek() !== Characters.RCurly
      ) {
        this.consumeWhitespace();
        const endedBlock = this.scanOptional(SymbolToken.RCurly);
        if (endedBlock) {
          yield endedBlock;
          checkRCurly = false;
          break;
        }
        yield* this.scanExpression();
      }
      if (checkRCurly) {
        yield this.scanRequired(SymbolToken.RCurly);
      }
    } else {
      yield* this.scanExpression();
    }
  }

  protected *scanFunction(): Iterable<Token> {
    yield this.scanIdentifier();
    this.consumeWhitespace();
    const lparen = this.scanOptional(SymbolToken.LParen);
    if (lparen) {
      yield lparen;
      let param = this.scanOptional(RegExpToken.Identifier);
      while (param) {
        yield param;
        param = this.scanOptional(RegExpToken.Identifier);
      }
      yield this.scanRequired(SymbolToken.RParen);
    }
    this.consumeWhitespace();
    yield this.scanRequired(StringToken.Arrow);
    yield* this.scanExpressionOrBlock();
  }

  protected scanIdentifier(): Token {
    const identifier = this.scanRequired(RegExpToken.Identifier);
    if (Lexer.keywords.has(identifier.value)) {
      return this.fail(`Invalid identifier: ${identifier.value} (Keyword)`);
    }
    return identifier;
  }

  protected scanLiteral(): Token | undefined {
    return (
      this.scanOptional(RegExpToken.LiteralBoolean) ||
      this.scanOptional(RegExpToken.LiteralNumber) ||
      this.scanOptional(RegExpToken.LiteralString)
    );
  }

  protected *scanParantheses(): Iterable<Token> {
    yield this.scanRequired(SymbolToken.LParen);
    while (!this.scanOptional(SymbolToken.RParen)) {
      yield* this.scanExpression();
    }
    yield new Token(SymbolToken.RParen, this.scanner.lastSpan(), ')');
  }
}
