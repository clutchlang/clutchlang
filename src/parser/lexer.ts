import { Codes, Scanner, Token, Tokens } from '../parser';

/**
 * Produces a stream of tokens from the @member scanner.
 *
 * Implements @type {Iterable<Token>}.
 */
export class Lexer implements Iterable<Token> {
  private static readonly matchIdentifier = /[_a-zA-Z][_a-zA-Z0-9]{0,30}/;
  private static readonly matchLiteralBool = /true|false/;
  private static readonly matchLiteralNumber = /-?\d+\.?\d*/;
  private static readonly matchLiteralString = /(?=["'])(?:"[^"\\]*(?:\\[\s\S][^"\\]*)*"|'[^'\\]*(?:\\[\s\S][^'\\]*)*')/;

  constructor(private readonly scanner: Scanner) {}

  private get substring() {
    let substring = this.scanner.contents.substring(this.scanner.position);
    const newLine = substring.indexOf('\n');
    if (newLine !== -1) {
      substring = substring.substring(0, newLine);
    }
    return substring;
  }

  public *[Symbol.iterator](): Iterator<Token> {
    this.consumeWhitespace();
    while (!this.scanner.isDone) {
      yield* this.scanCompilationUnit();
      this.consumeWhitespace();
    }
  }

  protected consumeWhitespace(): void {
    this.scanner.scan(/\s+/);
  }

  /**
   * Throws a terminal error at the current scanner position.
   *
   * @param message
   */
  protected fail(message: string): never {
    throw new SyntaxError(`${this.scanner.position}: ${message}`);
  }

  protected scanOptional(
    token: Tokens,
    pattern: number | string | RegExp
  ): Token | undefined {
    this.consumeWhitespace();
    const result = this.scanner.scan(pattern);
    return result ? new Token(token, this.scanner.lastMatch![0]) : undefined;
  }

  protected scanRequired(
    token: Tokens,
    pattern: number | string | RegExp
  ): Token {
    this.consumeWhitespace();
    const result = this.scanner.scan(pattern);
    if (!result) {
      this.fail(`Expected ${pattern}, got "${this.substring}"`);
    }
    return new Token(token, this.scanner.lastMatch![0]);
  }

  protected scanCompilationUnit(): Iterable<Token> {
    return this.scanFunction();
  }

  protected scanExpression(): Iterable<Token> {
    const expression =
      this.scanLiteral() ||
      this.scanOptional(Tokens.Identifier, Lexer.matchIdentifier);
    if (!expression) {
      if (this.scanner.peek() === Codes.LParen) {
        return this.scanParantheses();
      }
      return this.fail(`Expected expression, got "${this.substring}"`);
    }
    return [expression];
  }

  protected *scanExpressionOrBlock(): Iterable<Token> {
    this.consumeWhitespace();
    if (this.scanner.peek() === Codes.LCurly) {
      yield this.scanRequired(Tokens.LCurly, '{');
      while (!this.scanner.isDone && this.scanner.peek() !== Codes.RCurly) {
        this.consumeWhitespace();
        const endedBlock = this.scanOptional(Tokens.RCurly, '}');
        if (endedBlock) {
          yield endedBlock;
          break;
        }
        yield* this.scanExpression();
      }
    } else {
      yield* this.scanExpression();
    }
  }

  protected *scanFunction(): Iterable<Token> {
    yield this.scanIdentifier();
    yield this.scanRequired(Tokens.Arrow, '=>');
    yield* this.scanExpressionOrBlock();
  }

  protected scanIdentifier(): Token {
    return this.scanRequired(Tokens.Identifier, Lexer.matchIdentifier);
  }

  protected scanLiteral(): Token | undefined {
    return (
      this.scanOptional(Tokens.Boolean, Lexer.matchLiteralBool) ||
      this.scanOptional(Tokens.Number, Lexer.matchLiteralNumber) ||
      this.scanOptional(Tokens.String, Lexer.matchLiteralString)
    );
  }

  protected *scanParantheses(): Iterable<Token> {
    yield this.scanRequired(Tokens.LParen, '(');
    yield* this.scanExpression();
    yield this.scanRequired(Tokens.RParen, ')');
  }
}
