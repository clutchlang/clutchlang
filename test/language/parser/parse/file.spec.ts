import { tokenize } from '../../../../src/language/lexer';
import {
  ClutchParser,
  FileRoot,
  PrintTreeVisitor,
} from '../../../../src/language/parser';

describe('ClutchParser should parse statement', () => {
  function parseFile(file: string): FileRoot {
    const tokens = tokenize(file);
    return new ClutchParser(tokens).parseFileRoot();
  }

  it('should parse a basic program', () => {
    const file = parseFile(`
      main -> print('Hello World')
    `);
    expect(file.accept(new PrintTreeVisitor()).toString()).toMatchSnapshot();
  });

  it('should parse a basic program with a function with parameters', () => {
    const file = parseFile(`
      fib(n) -> if n <= 2 then fib(n - 1) else fib(n - 2)
    `);
    expect(file.accept(new PrintTreeVisitor()).toString()).toMatchSnapshot();
  });

  it('should parse a basic program with a constexpr function with parameters', () => {
    const file = parseFile(`
      const fib(n) -> if n <= 2 then fib(n - 1) else fib(n - 2)
    `);
    expect(file.accept(new PrintTreeVisitor()).toString()).toMatchSnapshot();
  });

  it('should parse a basic program with a function with statements', () => {
    const file = parseFile(`
      main -> {
        let message = 'Hello'
        print(message)
        return message
      }
    `);
    expect(file.accept(new PrintTreeVisitor()).toString()).toMatchSnapshot();
  });
});
