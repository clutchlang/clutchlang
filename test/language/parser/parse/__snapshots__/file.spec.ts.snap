// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClutchParser should parse statement should parse a basic program 1`] = `
"FileRoot:
  Elements:
    FunctionDeclaration:
      Name:
        SimpleName: main
      Parameters:
      Body:
        InvokeExpression:
          Target:
            SimpleName: print
          Parameters:
            LiteralString: 'Hello World'
"
`;

exports[`ClutchParser should parse statement should parse a basic program with a constexpr function with parameters 1`] = `
"FileRoot:
  Elements:
    FunctionDeclaration:
      Name:
        SimpleName: fib
      Const:
      Parameters:
        ParameterDeclaration:
          Name:
            SimpleName: n
      Body:
        ConditionalExpression:
          If:
            BinaryExpression:
              Left:
                SimpleName: n
              Operator: <=
              Right:
                LiteralNumber: 2
          Body:
            InvokeExpression:
              Target:
                SimpleName: fib
              Parameters:
                BinaryExpression:
                  Left:
                    SimpleName: n
                  Operator: -
                  Right:
                    LiteralNumber: 1
          Else:
            InvokeExpression:
              Target:
                SimpleName: fib
              Parameters:
                BinaryExpression:
                  Left:
                    SimpleName: n
                  Operator: -
                  Right:
                    LiteralNumber: 2
"
`;

exports[`ClutchParser should parse statement should parse a basic program with a function with parameters 1`] = `
"FileRoot:
  Elements:
    FunctionDeclaration:
      Name:
        SimpleName: fib
      Parameters:
        ParameterDeclaration:
          Name:
            SimpleName: n
          Type:
            SimpleName: Number
      Body:
        ConditionalExpression:
          If:
            BinaryExpression:
              Left:
                SimpleName: n
              Operator: <=
              Right:
                LiteralNumber: 2
          Body:
            InvokeExpression:
              Target:
                SimpleName: fib
              Parameters:
                BinaryExpression:
                  Left:
                    SimpleName: n
                  Operator: -
                  Right:
                    LiteralNumber: 1
          Else:
            InvokeExpression:
              Target:
                SimpleName: fib
              Parameters:
                BinaryExpression:
                  Left:
                    SimpleName: n
                  Operator: -
                  Right:
                    LiteralNumber: 2
"
`;

exports[`ClutchParser should parse statement should parse a basic program with a function with statements 1`] = `
"FileRoot:
  Elements:
    FunctionDeclaration:
      Name:
        SimpleName: main
      Parameters:
      Body:
        VariableDeclarationStatement:
          Name: message
          Assign:
            LiteralString: 'Hello'
        InvokeExpression:
          Target:
            SimpleName: print
          Parameters:
            SimpleName: message
        ReturnStatement:
          Return:
            SimpleName: message
"
`;
